Add RequireImportsSniff

sirbrillig opened the pull request:
This sniff requires that all functions, constants, and classes be imported explicitly and marks a warning if an unimported symbol is found.

That is, you cannot just import a namespace or assume that the symbols you use are in your namespace or the global namespace, each dependency must be explicitly imported. This makes dependency trees more clear.

For example:

```php
namespace Vehicles;
use function Vehicles\startCar;
class Car {
  public function drive() {
    startCar();
    goFaster(); // this will be a warning because `goFaster` was not imported
  }
}
```

Fixes #45 

Questions: 

- This will always add a warning for `define()` constants since they are never imported. Is there a way to tell the difference between a `const` constant and a `define` one? I don't think so.
- When used in WordPress this could be really annoying since it doesn't know any of the hundreds of WordPress symbols (`do_action()`, `is_wp_error()`, `WP_Error`, `mysql2date()`, etc.)
- Functions like `array_map()` [require the fully qualified namespace](https://blog.josephscott.org/2014/08/28/php-array_map-callback-expects-the-full-namespace/), so they're likely to trigger this warning a lot.

sirbrillig committed:
Tests: add test for RequireImportsSniff

sirbrillig committed:
RequireImportsSniff: implement sniff

sirbrillig commented:
@stuwest I'm pleased with how this turned out, but I have a feeling that no one will want to use it, particularly for WordPress. What do you think?

stuwest commented:
I agree a lot of WordPress devs won't love it, because they've gotten so used to just winging it with globals.

> This makes dependency trees more clear

But I think this is totally worth it.

This feels like a great item as a warning... but likely something that will take a lot of transitioning for existing code bases.

sirbrillig committed:
Tests: add failing test for importing global functions

sirbrillig committed:
SniffHelpers: skip import type in getImportNames

sirbrillig commented:
One thing I think that could be problematic here is when the namespace (or part of the namespace) serves to clarify an ambiguous symbol name.

For example, if there was the function `Calendar\update()` (arguably we should rename the function, but that could be a tall order for existing code bases), this sniff forces it to be called as `update()`, requiring the developer to look at the top of the file for the context, and possibly conflicting with other names.

I'm going to discuss this further with the Neutron team, but two paths I see going forward might be:

- Allow no import if using an explicit path (this nullifies a lot of the safety-checking ability of this sniff).
- Move this sniff to its own standard so it's easier to use and control.

sirbrillig committed:
Tests: add failing test for explicit namespace in call

sirbrillig committed:
Tests: add failing test for import alias

sirbrillig committed:
Tests: add failing test for alias in group import

sirbrillig committed:
SniffHelpers: refactor getImportNames to work with aliases

It now uses the last string before the end of the import (the semicolon,
comma, or closing curly brace).

sirbrillig committed:
RequireImportsSniff: disallow absolute symbol use

sirbrillig committed:
Tests: add test for imported but aliased function

sirbrillig committed:
Tests: add test for using absolute Exception class

sirbrillig committed:
RequireImportsSniff: include token in absolute use warning

sirbrillig committed:
Tests: add failing test for explicit namespace in catch

sirbrillig committed:
RequireImportsSniff: simplify hasNamespace

sirbrillig committed:
Tests: change catch test to custom class

sirbrillig committed:
RequireImportsSniff: allow isClass/isConstant to find namespaced symbols

sirbrillig commented:
Another issue I thought of is that functions like array_map() require the fully qualified namespace, so they're likely to trigger this warning a lot (added to the PR description). 😩

sirbrillig commented:
Closing this in favor of #59 which is simpler, less noisy, and hopefully more helpful.
